### PR TITLE
ReplyTo event feature

### PR DESCRIPTION
**Pull request by bmarty**

### CODE REVIEW

File: matrix-sdk/src/main/java/org/matrix/androidsdk/util/PermalinkUtils.java [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+import android.support.annotation.NonNull;
+import android.support.annotation.Nullable;
+import android.text.TextUtils;
+
+import org.matrix.androidsdk.rest.model.Event;
+
+/**
+ * Useful methods to create permalink
+ */
+public class PermalinkUtils {
```
Comment (ganfra) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/util/PermalinkUtils.java]: Maybe we should avoid adding more Utils classes and using static methods?
I do prefer having something as a PermalinkFactory we can  inject and mock in Unit Tests

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java [hunk 2]
```diff
@@ -2385,14 +2387,34 @@ public void onUnexpectedError(Exception e) {
         if (isEncrypted() && (null != mDataHandler.getCrypto())) {
             mDataHandler.updateEventState(event, Event.SentState.ENCRYPTING);
 
+            // Store the "m.relates_to" data and remove them from event content before encrypting the event content
+            final JsonElement relatesTo;
+
+            if (event.getContent().isJsonObject()
+                    && event.getContent().getAsJsonObject().has("m.relates_to")) {
```
Comment (ganfra) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: Seems like there is a getContentAsJsonObject method

Comment (bmarty) [reply matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: ok, fixed

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomMediaMessagesSender.java [hunk 3]
```diff
@@ -355,9 +367,178 @@ private Message buildTextMessage(RoomMediaMessage roomMediaMessage) {
...
+                    message.formatted_body = includeReplyToToFormattedBody(replyToEvent,
+                            replyToFormattedBody,
+                            replyToEventIsAlreadyAReply,
+                            replyContent,
+                            msgType.equals(Message.MSGTYPE_EMOTE));
+
+                    // Note: we need to force the format to Message.FORMAT_MATRIX_HTML
+                    message.format = Message.FORMAT_MATRIX_HTML;
+                } else {
+                    // Ensure there will not be "m.relates_to" data in the sent event
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomMediaMessagesSender.java]: TODO: add error log for developers

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomMediaMessagesSender.java [hunk 4]
```diff
@@ -355,9 +367,178 @@ private Message buildTextMessage(RoomMediaMessage roomMediaMessage) {
...
+                    roomMediaMessage.setReplyToEvent(null);
+                }
+            } else {
+                // Ensure there will not be "m.relates_to" data in the sent event
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/RoomMediaMessagesSender.java]: TODO: add error log for developers

File: matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java [hunk 5]
```diff
@@ -2838,45 +2862,90 @@ private void initRoomMediaMessagesSender() {
...
+    /**
+     * Indicate if replying to the provided event is supported.
+     * Only event of type Event.EVENT_TYPE_MESSAGE are supported for the moment
+     *
+     * @param replyToEvent the event to reply to
+     * @return true if it is possible to reply to this event
+     */
+    public boolean canReplyTo(@Nullable Event replyToEvent) {
+        return replyToEvent != null
+                && Event.EVENT_TYPE_MESSAGE.equals(replyToEvent.getType());
```
Comment (bmarty) [inline matrix-sdk/src/main/java/org/matrix/androidsdk/data/Room.java]: check msgtype as well?

**Pull request merged**